fetch.bs

```diff
@@ -3225,7 +3238,112 @@ through TLS using ALPN. The protocol cannot be spoofed through HTTP requests in
+      </li>
+    </ul>
+  </li>
+
+  <li><p>If <var>request</var>'s <code>https-upgrade-fallback-url</code> is non-null, this is a
+  fallback request that shouldn't be upgraded again. Clear the <code>https-upgrade</code> and
+  <code>https-upgrade-fallback-url</code> flags and return <var>request</var>.</li>
+
+  <li><p>Otherwise, upgrade the request by setting the following fields:
+    <ul>
+      <li>Set <code>https-upgrade-fallback-url</code> to request's URL</li>
+      <li>Set <var>request</var>'s URL's scheme to <code>"https"</code></li>
+      <li>Set <code>https-upgrade</code> flag to true</li>
+    </ul>
+  </li>
+</ol>
+
+<h4 id=https-upgrades-fallback>Fallback algorithm</h4>
+
+<p>Runs after the Upgrade algorithm. This algorithm determines whether the upgraded
+request completed without a network error. If not, it retries the request over <code>"http"</code>.
+
+<ol>
+  <li><p>If the <code>https-upgrade</code> flag is not set, return without modifying request.</li>
+
+  <li><p>Otherwise, this is an upgraded request. Let <var>upgradedResponse</var> be the result of
+  fetching the upgraded request.</li>
+
+  <li>
+    <p>If <var>upgradedResponse</var> is a network error, initiate a fallback http load by doing the
+    following:
+
+    <ul>
+      <li><p>Set <var>request</var>'s URL to the value of <var>request</var>'s
+        <code>https-upgrade-fallback-url</code> flag</li>
+
+      <li><p>Set response to the result of running <a>HTTP-redirect fetch</a> given fetchParams and
+        <var>upgradedResponse</var>.</li>
+    </ul>
+  </li>
+
+  <li><p>Otherwise, the upgrade was successful. Return <var>upgradedResponse</var>.</li>
+</ol>
+
+<p>
+Optionally, user agents may implement a fast-fallback option by canceling slow fetches on upgraded
+requests and quickly initiate a fallback http load. 
+
+</div>
+
+<h4>Examples</h4>
+
+<div id=example-https-upgrade-good-https class=example>
+ <p>
+ <code>a.com</code> serves both <code>http://a.com</code> and <code>https://a.com</code>.
+ An eligible request to <code>http://a.com</code> will be upgraded to <code>https://a.com</code>. 
+</div>
+
+<div id=example-https-upgrade-bad-https class=example>
+ <p>
+  <code>a.com</code> serves <code>http://a.com</code> but refuses connections on
+  <code>https://a.com</code>.
+  An eligible request to <code>http://a.com</code> will be upgraded to <code>https://a.com</code>,
+  but the fetch will fail.
+  A fallback request will be initiated to <code>http://a.com</code>.
+</div>
 
+<div id=example-https-upgrade-slow-https class=example>
+ <p>
+  Optional fast-fallback example: <code>a.com</code> serves <code>http://a.com</code> and loads very
+  slowly over <code>https://a.com</code>.
+  An eligible request to <code>http://a.com</code> will be upgraded to <code>https://a.com</code>.
+  If the upgraded request doesn't return a response for N seconds, the fetch will be canceled.
```

Comment:
I think it might be better to either explicitly make this implementation-defined, or to cut this entirely as it isn't particularly interesting for the spec IMO, as implementations can already trigger a network error for any number of reasons -- this is just a callout that Chrome specifically caps the network-level timeout to a shorter time than it normally does for this feature.